### PR TITLE
Store Video progress in client side cookie

### PR DESCRIPTION
**Pull request by mehalshah**

### CODE REVIEW

**Comment by philbogle** [comment]
Looking (will finish on the bus)

File: dashboard/app/assets/javascripts/sendReport.js [hunk 1]
```diff
@@ -36,7 +36,7 @@ var sendReport = function(report) {
   var queryString = queryItems.join('&');
 
   dashboard.clientState.trackProgress(report.result, report.lines, report.testResult, appOptions.level.scriptLevelId);
-
+  console.log(report.callback);
```
Comment (philbogle) [inline dashboard/app/assets/javascripts/sendReport.js]: I think this can go away.

File: dashboard/app/assets/javascripts/videos.js.erb [hunk 2]
```diff
@@ -24,6 +24,11 @@ function createVideo(options) {
 //   name - a string.
 //   redirect - the redirect page after the video is dismissed.
 function showVideoDialog(options) {
+
+  if (dashboard.clientState.hasSeenVideo(options.key)) {
+    return ;
```
Comment (philbogle) [inline dashboard/app/assets/javascripts/videos.js.erb]: Delete space before ;

File: dashboard/app/assets/javascripts/videos.js.erb [hunk 3]
```diff
@@ -45,6 +50,11 @@ function showVideoDialog(options) {
   $div = $(dialog.div);
   $div.addClass('video-modal');
 
+  $('.video-modal #x-close').click(function () {
+    console.log('Marking ' + options.key + ' as seen');
```
Comment (philbogle) [inline dashboard/app/assets/javascripts/videos.js.erb]: Log spam not needed.

File: dashboard/app/assets/javascripts/videos.js.erb [hunk 4]
```diff
@@ -45,6 +50,11 @@ function showVideoDialog(options) {
   $div = $(dialog.div);
   $div.addClass('video-modal');
 
+  $('.video-modal #x-close').click(function () {
```
Comment (philbogle) [inline dashboard/app/assets/javascripts/videos.js.erb]: I think there are other ways to get rid of the video that don't follow this code path, for example hitting escape or clicking on another bubble.

Perhaps we want to consider the video seen as soon as the user looks at the page and the video is presented?  (Could double check with Brendan or Alice if you aren't sure.)

**Comment by philbogle** [comment]
LGTM after comments are addressed.  The only one of substance concerns when we consider the video seen.  I believe its closest to current behavior to consider the video seen as soon as it starts playing, even if the user doesn't click the X.

(Note in the subsequent change that when you migrate cookie state you'll need to merge the client and server state on the client.)

File: dashboard/app/assets/javascripts/sendReport.js (hunk 1 above)
Comment (mehalshah) [inline dashboard/app/assets/javascripts/sendReport.js]: Removed

File: dashboard/app/assets/javascripts/videos.js.erb (hunk 2 above)
Comment (mehalshah) [inline dashboard/app/assets/javascripts/videos.js.erb]: Removed

File: dashboard/app/assets/javascripts/videos.js.erb (hunk 3 above)
Comment (mehalshah) [inline dashboard/app/assets/javascripts/videos.js.erb]: Removed

File: dashboard/app/assets/javascripts/videos.js.erb (hunk 4 above)
Comment (mehalshah) [inline dashboard/app/assets/javascripts/videos.js.erb]: I can do it when the dialog is removed. Handles other ways of escaping.

**Comment by laurelfan** [comment]
I think current behavior is to consider the video seen as soon as the page
containing the autoplay video is loaded (or more precisely, as soon as the
browser deals with the Set-Cookie in the http response for the page),
whether not the video or dialog itself has loaded.

In practice (other than some edge cases around the js on the page being
broken) this means that noautoplay=1 will work differently (this param is
only used by tests and not by real users). I have no opinion on which
behavior is better (and don't think it matters that much), but you may have
to fix some ui tests if you change it.

**Pull request merged**